File: kotlin/src/main/kotlin/fundamentals/Operators.kt
package fundamentals

data class Point(var x: Int, var y: Int) {
    // Operator overloading example
    operator fun plus(other: Point) = Point(x + other.x, y + other.y)
    operator fun minus(other: Point) = Point(x - other.x, y - other.y)
    operator fun unaryMinus() = Point(-x, -y)
}

class Operators {
    fun demonstrate() {
        println("=== Operators ===\n")

        // Arithmetic operators
        println("1. Arithmetic Operators:")
        val a = 10
        val b = 3
        println("a = $a, b = $b")
        println("Addition (a + b): ${a + b}")
        println("Subtraction (a - b): ${a - b}")
        println("Multiplication (a * b): ${a * b}")
        println("Division (a / b): ${a / b}")
        println("Modulus (a % b): ${a % b}")

        // Integer vs Float division
        println("\n2. Integer vs Float Division:")
        println("Integer division (10 / 3): ${10 / 3}")
        println("Float division (10.0 / 3.0): ${10.0 / 3.0}")

        // Increment and decrement
        println("\n3. Increment and Decrement:")
        var x = 5
        println("x = $x")
        println("++x = ${++x}")  // Prefix increment
        println("x++ = ${x++}")  // Postfix increment
        println("x = $x")
        println("--x = ${--x}")  // Prefix decrement
        println("x-- = ${x--}")  // Postfix decrement
        println("x = $x")

        // Assignment operators
        println("\n4. Assignment Operators:")
        var value = 10
        println("Initial value: $value")
        value += 5
        println("After += 5: $value")
        value -= 3
        println("After -= 3: $value")
        value *= 2
        println("After *= 2: $value")
        value /= 4
        println("After /= 4: $value")
        value %= 3
        println("After %= 3: $value")

        // Comparison operators
        println("\n5. Comparison Operators:")
        val num1 = 10
        val num2 = 20
        println("num1 = $num1, num2 = $num2")
        println("num1 > num2: ${num1 > num2}")
        println("num1 < num2: ${num1 < num2}")
        println("num1 >= num2: ${num1 >= num2}")
        println("num1 <= num2: ${num1 <= num2}")
        println("num1 == num2: ${num1 == num2}")
        println("num1 != num2: ${num1 != num2}")

        // Logical operators
        println("\n6. Logical Operators:")
        val t = true
        val f = false
        println("t = $t, f = $f")
        println("AND (t && f): ${t && f}")
        println("OR (t || f): ${t || f}")
        println("NOT (!t): ${!t}")

        // Bitwise operators
        println("\n7. Bitwise Operators:")
        val bits1 = 0b1100  // 12 in binary
        val bits2 = 0b1010  // 10 in binary
        println("bits1 = ${bits1.toString(2).padStart(4, '0')}")
        println("bits2 = ${bits2.toString(2).padStart(4, '0')}")
        println("AND (bits1 and bits2): ${(bits1 and bits2).toString(2).padStart(4, '0')}")
        println("OR (bits1 or bits2): ${(bits1 or bits2).toString(2).padStart(4, '0')}")
        println("XOR (bits1 xor bits2): ${(bits1 xor bits2).toString(2).padStart(4, '0')}")
        println("Inverse of bits1: ${bits1.inv() and 0b1111}")
        println("Left shift (bits1 shl 1): ${(bits1 shl 1).toString(2).padStart(4, '0')}")
        println("Right shift (bits1 shr 1): ${(bits1 shr 1).toString(2).padStart(4, '0')}")

        // Operator overloading
        println("\n8. Operator Overloading:")
        val p1 = Point(2, 3)
        val p2 = Point(3, 4)
        println("p1 = $p1")
        println("p2 = $p2")
        println("p1 + p2 = ${p1 + p2}")
        println("p1 - p2 = ${p1 - p2}")
        println("-p1 = ${-p1}")

        // Range operator
        println("\n9. Range Operator:")
        val range = 1..5
        println("Range 1..5: $range")
        println("4 in range: ${4 in range}")
        println("6 in range: ${6 in range}")
        
        // Safe call operator
        println("\n10. Safe Call Operator:")
        val nullableString: String? = null
        println("Length of nullable string: ${nullableString?.length}")
        
        // Elvis operator
        println("\n11. Elvis Operator:")
        val length = nullableString?.length ?: -1
        println("Length with Elvis operator: $length")

        println() // Empty line
    }
}

File: scala/src/main/scala/fundamentals/Operators.scala
package fundamentals

object Operators:
  def demonstrate(): Unit =
    println("=== Operators ===\n")
    
    // Arithmetic operators
    println("1. Arithmetic Operators:")
    println(s"Addition: 5 + 3 = ${5 + 3}")
    println(s"Subtraction: 5 - 3 = ${5 - 3}")
    println(s"Multiplication: 5 * 3 = ${5 * 3}")
    println(s"Division: 5 / 2 = ${5 / 2}")  // Integer division
    println(s"Division (float): 5.0 / 2 = ${5.0 / 2}")
    println(s"Modulus: 5 % 3 = ${5 % 3}")
    println(s"Power: math.pow(2, 3) = ${math.pow(2, 3)}")
    
    // Unary operators
    println("\n2. Unary Operators:")
    var x = 10
    println(s"Original x: $x")
    println(s"Unary minus -x: ${-x}")
    println(s"Unary plus +x: ${+x}")
    println(s"Prefix increment: ${x += 1; x}")  // Returns 11
    println(s"Prefix decrement: ${x -= 1; x}")  // Returns 10
    
    // Comparison operators
    println("\n3. Comparison Operators:")
    val a = 5
    val b = 3
    println(s"a = $a, b = $b")
    println(s"Equal to: a == b is ${a == b}")
    println(s"Not equal to: a != b is ${a != b}")
    println(s"Greater than: a > b is ${a > b}")
    println(s"Less than: a < b is ${a < b}")
    println(s"Greater than or equal to: a >= b is ${a >= b}")
    println(s"Less than or equal to: a <= b is ${a <= b}")
    
    // Logical operators
    println("\n4. Logical Operators:")
    val p = true
    val q = false
    println(s"p = $p, q = $q")
    println(s"AND: p && q is ${p && q}")
    println(s"OR: p || q is ${p || q}")
    println(s"NOT: !p is ${!p}")
    
    // Bitwise operators
    println("\n5. Bitwise Operators:")
    val m = 60  // 111100
    val n = 13  // 001101
    println(s"m = $m (${m.toBinaryString}), n = $n (${n.toBinaryString})")
    println(s"Bitwise AND: m & n = ${m & n} (${(m & n).toBinaryString})")
    println(s"Bitwise OR: m | n = ${m | n} (${(m | n).toBinaryString})")
    println(s"Bitwise XOR: m ^ n = ${m ^ n} (${(m ^ n).toBinaryString})")
    println(s"Bitwise NOT: ~m = ${~m} (${(~m).toBinaryString})")
    println(s"Left shift: m << 2 = ${m << 2} (${(m << 2).toBinaryString})")
    println(s"Right shift: m >> 2 = ${m >> 2} (${(m >> 2).toBinaryString})")
    println(s"Unsigned right shift: m >>> 2 = ${m >>> 2} (${(m >>> 2).toBinaryString})")
    
    // Assignment operators
    println("\n6. Assignment Operators:")
    var value = 10
    println(s"Original value: $value")
    value += 5; println(s"After += 5: $value")
    value -= 3; println(s"After -= 3: $value")
    value *= 2; println(s"After *= 2: $value")
    value /= 4; println(s"After /= 4: $value")
    value %= 3; println(s"After %= 3: $value")
    
    // Pattern matching operators
    println("\n7. Pattern Matching:")
    val number = 42
    val result = number match
      case 0 => "Zero"
      case n if n > 0 => "Positive"
      case _ => "Negative"
    println(s"Pattern matching result for $number: $result")
    
    // Type test operators
    println("\n8. Type Test Operators:")
    val obj: Any = "Hello"
    println(s"obj.isInstanceOf[String]: ${obj.isInstanceOf[String]}")
    println(s"obj.isInstanceOf[Int]: ${obj.isInstanceOf[Int]}")
    println(s"obj.asInstanceOf[String].length: ${obj.asInstanceOf[String].length}")
    
    println() // Empty line